## 4.6.1 Combining & Exporting Data

# Merging Prepared Instacart orders data with the new orders_products_prior

In [4]:
#import libraries

In [70]:
import pandas as pd
import numpy as np
import os

### Import the data sets into Jupyter

In [72]:
path = r"C:\Users\hp\08-08-2024 Instacart Basket Analysis"

In [74]:
# Import dataset orders_products_prior.csv
df_ords_prods_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

In [75]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked_4.5.csv'), index_col = False)

In [76]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   order_dow               int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


In [77]:
#check the size and datatype of the dataframe columns
df_ords_prods_prior.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 989.8 MB


### Checking dimensions of the dataframe

In [78]:
df_ords_prods_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [86]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329        1    prior             1          2   
1           1   2398795        1    prior             2          3   
2           2    473747        1    prior             3          3   
3           3   2254736        1    prior             4          4   
4           4    431534        1    prior             5          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     0.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [18]:
df_ords_prods_prior.shape

(32434489, 4)

In [88]:
df_ords.shape

(3421083, 8)

In [92]:
# Dropping unnecessary columns to reduce the file size
df_ords = df_ords.drop(['eval_set', 'Unnamed: 0'], axis=1)


In [94]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [96]:
df_ords.shape

(3421083, 6)

In [98]:
#changing the data type to reduce the size of the file
df_ords_prods_comb = df_ords_prods_prior.astype({
    'order_id': 'int32',
    'product_id': 'int32',
    'add_to_cart_order': 'int8',
    'reordered': 'int8',
    })

In [100]:
df_ords_prods_comb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int32
 1   product_id         int32
 2   add_to_cart_order  int8 
 3   reordered          int8 
dtypes: int32(2), int8(2)
memory usage: 309.3 MB


# Above shows that memory usage has reduced to 309MB from 990 MB

In [102]:
#merge the dataframes
df_merged_large = df_ords.merge(df_ords_prods_comb, on = 'order_id', indicator = True)

In [103]:
df_merged_large.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     0.0         196                  1          0   both  
1                     0.0       14084                  2          0   both  
2                     0.0       12427                  3          0   both  
3                     0.0       26088                  4          0   both  
4                     0.0       26405                  5          0   both

In [114]:
#checking if its a full match
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [118]:
df_merged_large.shape

(32434489, 10)

## Exporting the merged dataframe

In [120]:
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))

In [122]:
df_merged_large.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2539329        1             1          2                  8   
2   2539329        1             1          2                  8   
3   2539329        1             1          2                  8   
4   2539329        1             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     0.0         196                  1          0   both  
1                     0.0       14084                  2          0   both  
2                     0.0       12427                  3          0   both  
3                     0.0       26088                  4          0   both  
4                     0.0       26405                  5          0   both